### PR TITLE
Routing import memory optimizations

### PR DESCRIPTION
**Pull request by mkurc-ant**
This PR adds optimizations to the `prjxray_routing_import.py` functionality allowing to significantly reduce its memory footprint. 

Before the optimization import of routing for the whole 50T rr graph took **18.5 GB** and **336 sec**. Once they are applied it takes only **3.6 GB** and **397 sec** (the db read from a HDD drive).

The optimizations include:

- Removal of loading of the whole rr graph to memory at once in favor of parsing it iteratively. Also the data is not stored in ET.Element objects which are pretty memory consuming.
- Refactored rr graph serialization so the data is directly written into the XML file rather than being stored in ET.Element objects and then serialized.
- Removal of SQL database cache. The database is read from the file directly.

This PR also adds a memory usage logging utility script.

### CODE REVIEW

**Comment by litghost** [comment]
Testarch needs an update:
```
Generating rr_graph_10x10_dummy.rr_graph.real.xml
Importing input g
Traceback (most recent call last):
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-testarch/testarch/utils/testarch_graph.py", line 400, in <module>
    main()
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-testarch/testarch/utils/testarch_graph.py", line 395, in main
    verbose=args.verbose
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-testarch/testarch/utils/testarch_graph.py", line 342, in rebuild_graph
    output_file_name=fn_out,
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-testarch/utils/lib/rr_graph_xml/graph2.py", line 256, in __init__
    graph_input = graph_from_xml(input_file_name, progressbar)
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-testarch/utils/lib/rr_graph_xml/graph2.py", line 68, in graph_from_xml
    for path, element in progressbar(iterate_xml(input_file_name)):
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-testarch/utils/lib/rr_graph_xml/graph2.py", line 29, in iterate_xml
    doc = ET.iterparse(xml_file, events=('start', 'end'))
  File "src/lxml/iterparse.pxi", line 78, in lxml.etree.iterparse.__init__
TypeError: expected str, bytes or os.PathLike object, not lxml.etree._ElementTree
testarch/devices/CMakeFiles/file_testarch_devices_rr_graph_10x10_dummy.rr_graph.real.xml.dir/build.make:62: recipe for target 'testarch/devices/rr_graph_10x10_dummy.rr_graph.real.xml' failed
make[3]: *** [testarch/devices/rr_graph_10x10_dummy.rr_graph.real.xml] Error 1
CMakeFiles/Makefile2:199395: recipe for target 'testarch/devices/CMakeFiles/file_testarch_devices_rr_graph_10x10_dummy.rr_graph.real.xml.dir/all' failed
make[2]: *** [testarch/devices/CMakeFiles/file_testarch_devices_rr_graph_10x10_dummy.rr_graph.real.xml.dir/all] Error 2
CMakeFiles/Makefile2:75: recipe for target 'CMakeFiles/all_testarch.dir/rule' failed
make[1]: *** [CMakeFiles/all_testarch.dir/rule] Error 2
```

File: xc7/utils/prjxray_routing_import.py [hunk 1]
```diff
@@ -1000,6 +1003,10 @@ def main():
 
     args = parser.parse_args()
 
+    # Initialize the memory log
+    memlog = MemoryLog(args.write_rr_graph + ".memory.log")
```
Comment (litghost) [inline xc7/utils/prjxray_routing_import.py]: We can get rid of these calls to MemoryLog now

File: xc7/utils/prjxray_routing_import.py [hunk 2]
```diff
@@ -36,8 +36,10 @@
 import re
 import functools
 import pickle
+import time
```
Comment (litghost) [inline xc7/utils/prjxray_routing_import.py]: Unused import

File: xc7/utils/prjxray_routing_import.py (hunk 1 above)
Comment (mkurc-ant) [reply xc7/utils/prjxray_routing_import.py]: Removed.

File: xc7/utils/prjxray_routing_import.py (hunk 2 above)
Comment (mkurc-ant) [reply xc7/utils/prjxray_routing_import.py]: Removed.

**Comment by mkurc-ant** [comment]
I fixed the testarch. Let's see what the CI has to say

**Comment by litghost** [comment]
Another error in fasm2bels:
```
[1385/1917] Generating 5-lut_cascade_1_dummy_artix7_xc7a50t-basys3_test/artix7-xc7a50t-basys3-roi-virt-xc7a50t-basys3-test/top.route
FAILED: cd /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/srl/srl16_init && /usr/bin/cmake -E env PYTHONPATH=/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/env/conda/lib/python3.7/site-packages:/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray:/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/third_party/fasm:/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/xc7:/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/utils /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/env/conda/bin/python3 -mfasm2bels --part xc7a35tcpg236-1 --connection_database /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/archs/artix7/devices/xc7a50t-basys3-roi-virt/channels.db --db_root /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray-db/artix7 --rr_graph /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/archs/artix7/devices/rr_graph_xc7a50t-basys3_test.rr_graph.virt.xml --route /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/srl/srl16_init/srl16_init/artix7-xc7a50t-basys3-roi-virt-xc7a50t-basys3-test/top.route --iostandard_defs /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/srl/srl16_init/srl16_init/artix7-xc7a50t-basys3-roi-virt-xc7a50t-basys3-test/top.eblif.iostandard.json --bitread /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/third_party/prjxray/tools/bitread --bit_file /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/srl/srl16_init/srl16_init/artix7-xc7a50t-basys3-roi-virt-xc7a50t-basys3-test/top.bit --fasm_file /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/srl/srl16_init/srl16_init/artix7-xc7a50t-basys3-roi-virt-xc7a50t-basys3-test/top.bit.fasm --pcf /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/common/basys3.pcf --eblif /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/srl/srl16_init/srl16_init/artix7-xc7a50t-basys3-roi-virt-xc7a50t-basys3-test/top.eblif --top top /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/srl/srl16_init/srl16_init/artix7-xc7a50t-basys3-roi-virt-xc7a50t-basys3-test/top_bit.v /tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/xc7/tests/srl/srl16_init/srl16_init/artix7-xc7a50t-basys3-roi-virt-xc7a50t-basys3-test/top_bit.v.tcl
Traceback (most recent call last):
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/env/conda/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/build/env/conda/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/xc7/fasm2bels/__main__.py", line 4, in <module>
    main()
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/xc7/fasm2bels/fasm2bels.py", line 256, in main
    net_map = load_net_list(conn, args.rr_graph, args.route_file)
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/xc7/fasm2bels/fasm2bels.py", line 166, in load_net_list
    read_xml_file(rr_graph_file), build_pin_edges=False
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/utils/lib/rr_graph_xml/graph2.py", line 256, in __init__
    graph_input = graph_from_xml(input_file_name, progressbar)
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/utils/lib/rr_graph_xml/graph2.py", line 68, in graph_from_xml
    for path, element in progressbar(iterate_xml(input_file_name)):
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/utils/lib/rr_graph_xml/graph2.py", line 29, in iterate_xml
    doc = ET.iterparse(xml_file, events=('start', 'end'))
  File "src/lxml/iterparse.pxi", line 78, in lxml.etree.iterparse.__init__
TypeError: expected str, bytes or os.PathLike object, not lxml.etree._ElementTree
```

**Comment by mkurc-ant** [comment]
Fixed all the errors, success!

**Pull request merged**